### PR TITLE
feat(compass-collection): Output Validation Followup - Mock Data Generator LLM CLOUDP-347048

### PR DESCRIPTION
**Pull request by ncarbon**
## Description
This pull request introduces more validation and rendering improvements for handling faker method arguments in the mock data generator modal. The main changes include adding comprehensive validation logic for faker arguments and method names, updating UI components to display faker arguments, and refactoring validation logic out of the module to a shared utility. 

Screenshot 2025-09-25 at 10 55 11 AM
Screenshot 2025-09-25 at 1 55 11 PM

Screenshot 2025-09-25 at 1 03 48 PM

**Validation and Utility Enhancements:**

- Added `areFakerArgsValid` and improved `isValidFakerMethod` utility functions in `utils.ts` to thoroughly validate faker arguments (including nested and object types) and method names, enforcing limits on argument length, string size, and value types.
- Refactored the schema validation logic in `collection-tab.ts` to use the new `isValidFakerMethod` utility, ensuring faker method and argument validation is consistent and centralized. [[1]](diffhunk://#diff-ec5a026a2b50a34c6e78a5b6e90b526872f6e0e5320fcf6696dae3405c98dff5L44-R44) [[2]](diffhunk://#diff-ec5a026a2b50a34c6e78a5b6e90b526872f6e0e5320fcf6696dae3405c98dff5L719-L750) [[3]](diffhunk://#diff-ec5a026a2b50a34c6e78a5b6e90b526872f6e0e5320fcf6696dae3405c98dff5L778-R754)

**Type and Interface Updates:**

- Extended the `FakerArg` type to support nested arrays, enabling more flexible representation of faker arguments.

**UI and Rendering Improvements:**

- Updated `FakerMappingSelector` to render `TextInput` components for each faker argument, supporting various types (string, number, boolean, arrays, and objects) and displaying them in a read-only format. [[1]](diffhunk://#diff-e52fc6b570f9926779f98cab47885656ae6dd366fb9c2a95fd8b8e0f71638a6bR10-R15) [[2]](diffhunk://#diff-e52fc6b570f9926779f98cab47885656ae6dd366fb9c2a95fd8b8e0f71638a6bR29-R128) [[3]](diffhunk://#diff-e52fc6b570f9926779f98cab47885656ae6dd366fb9c2a95fd8b8e0f71638a6bL75-R167)
- Updated `FakerSchemaEditorContent` to pass the active faker arguments to the mapping selector for display. [[1]](diffhunk://#diff-9cb85785758b0ca96f24359491cf85b9ea44159af60fb4062d32675984b6045dR67) [[2]](diffhunk://#diff-9cb85785758b0ca96f24359491cf85b9ea44159af60fb4062d32675984b6045dR113)

**Testing:**

- Added comprehensive unit tests for `areFakerArgsValid` and `isValidFakerMethod` in `utils.spec.ts`, covering edge cases such as deeply nested structures, invalid argument types, and method existence.
- Added an integration test to ensure that excessively large faker arguments are not displayed in the modal.

### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] If this change updates the UI, screenshots/videos are added and a design review is requested
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.spec.tsx [hunk 1]
```diff
@@ -62,6 +62,7 @@ describe('MockDataGeneratorModal', () => {
       },
     };
 
+    // @ts-expect-error // TODO: fix ts error
```
Comment (ncarbon) [inline packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.spec.tsx]: Fixing this

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+      });
+
+      it('returns true for valid method with no args if args are invalid but fallback works', () => {
+        // internet.email does not take arguments, so passing one should fallback to []
+        const result = isValidFakerMethod('internet.email', []);
+        expect(result.isValid).to.be.true;
+        expect(result.fakerArgs).to.deep.equal([]);
+      });
+
+      it('returns false for valid method with invalid arguments and fallback fails', () => {
```
Comment (jcobis) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.spec.ts]: Shouldn't this be `invalid arguments with valid method falls back to true and strips args`? That's what's happening I think, which is why this test is failing right now

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+      it('returns false for helpers.arrayElement with invalid arguments', () => {
+        // Exceeding max args length
+        const arr = Array(11).fill('x');
+        const result = isValidFakerMethod('helpers.arrayElement', [arr]);
+        expect(result.isValid).to.be.false;
+        expect(result.fakerArgs).to.deep.equal([]);
+      });
+
+      it('returns false for method with invalid fakerArgs', () => {
```
Comment (jcobis) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.spec.ts]: [My comment](https://github.com/mongodb-js/compass/pull/7365/files#r2373235731) applies here too I believe. Both tests should expect `isValid: true` and `fakerArgs: []`

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      try {
+        callable();
+        return { isValid: true, fakerArgs: [] };
+      } catch {
+        // Intentionally ignored: method is invalid without arguments as well.
+        return { isValid: false, fakerArgs: [] };
+      }
+    }
+  }
+  return { isValid: false, fakerArgs: [] };
```
Comment (jcobis) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: What if we refactored a bit for clarity:

```suggestion
  // If args are present and safe, try calling with args
  if (args.length > 0 && areFakerArgsValid(args)) {
    try {
      callable(...args);
      return { isValid: true, fakerArgs: args };
    } catch {
      // Call with args failed. Fall through to trying without args
    }
  }

  // Try without args (either because args were invalid or args failed)
  try {
    callable();
    return { isValid: true, fakerArgs: [] };
  } catch {
    // Method doesn't work even without args
    return { isValid: false, fakerArgs: [] };
  }
```

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        return false;
+      }
+    } else if (typeof arg === 'string') {
+      if (arg.length > MAX_FAKER_STRING_LENGTH) {
+        return false;
+      }
+    } else if (typeof arg === 'boolean') {
+      // booleans are always valid, continue
+    } else if (Array.isArray(arg)) {
+      if (!areFakerArgsValid(arg)) {
```
Comment (jcobis) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: Should we limit the depth of this to prevent potential infinite recursion? We could pass a depth param and limit to say, 3 levels

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.spec.ts (hunk 2 above)
Comment (ncarbon) [reply packages/compass-collection/src/components/mock-data-generator-modal/utils.spec.ts]: Yep, forgot to update it.

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+) {
+  try {
+    if (areFakerArgsValid(args)) {
+      callable(...args);
+      return { isValid: true, fakerArgs: args };
+    } else {
+      callable();
+      return { isValid: true, fakerArgs: [] };
+    }
+  } catch {
```
Comment (kpamaran) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: A warning when there's an exception or the args are invalid would be valuable for fine tuning the prompt

File: packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.spec.tsx [hunk 7]
```diff
@@ -538,6 +539,44 @@ describe('MockDataGeneratorModal', () => {
...
+              probability: 1.0,
+            },
+          },
+          sampleDocument: { name: 'Peaches' },
+        },
+      });
+
+      // advance to the schema editor step
+      userEvent.click(screen.getByText('Confirm'));
+      await waitFor(() => {
```
Comment (kpamaran) [inline packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.spec.tsx:630]: seems to be an assertion missing? I'm not connecting how "faker args that are too large" are not shown

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx [hunk 8]
```diff
@@ -24,16 +26,106 @@ const labelStyles = css({
   fontWeight: 600,
 });
 
+/**
+ * Renders read-only TextInput components for each key-value pair in a faker arguments object.
+ */
+const getFakerArgsInputFromObject = (fakerArgsObject: Record<string, any>) => {
```
Comment (kpamaran) [inline packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: nit: use render in name to clarify React Elements are returned, something like `renderFakerArgsInput`

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx [hunk 9]
```diff
@@ -24,16 +26,106 @@ const labelStyles = css({
...
+ * Renders TextInput components for each faker argument based on its type.
+ */
+const getFakerArgsInput = (fakerArgs: FakerArg[]) => {
+  return fakerArgs.map((arg, idx) => {
+    if (typeof arg === 'string' || typeof arg === 'boolean') {
+      return (
+        <TextInput
+          key={`faker-arg-${idx}`}
+          type="text"
+          label="Faker Arg"
```
Comment (kpamaran) [inline packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: labels should be unique across the form's inputs

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx [hunk 10]
```diff
@@ -24,16 +26,106 @@ const labelStyles = css({
   fontWeight: 600,
 });
 
+/**
+ * Renders read-only TextInput components for each key-value pair in a faker arguments object.
+ */
+const getFakerArgsInputFromObject = (fakerArgsObject: Record<string, any>) => {
+  return Object.entries(fakerArgsObject).map(([key, item]: [string, any]) => {
+    if (typeof item === 'string' || typeof item === 'boolean') {
```
Comment (kpamaran) [inline packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: shouldn't `boolean` use something like a radio option instead of arbitrary text?

**Comment by kpamaran** [comment]
The faker arg form experience itself is interesting, I wonder how well users can intuit that the args are positional. Could be a follow-up. Maybe the user can be shown an inline preview of the faker call (cc @jcobis), right now they'd need to make the connection by reading the Script output

**Comment by ncarbon** [comment]
> The faker arg form experience itself is interesting, I wonder how well users can intuit that the args are positional. Could be a follow-up. Maybe the user can be shown an inline preview of the faker call (cc @jcobis), right now they'd need to make the connection by reading the Script output

Yeah, good question. If we can't show a specific label for each faker argument, an inline preview of the faker call would be good UI to show.

**Comment by jcobis** [comment]
> > The faker arg form experience itself is interesting, I wonder how well users can intuit that the args are positional. Could be a follow-up. Maybe the user can be shown an inline preview of the faker call (cc @jcobis), right now they'd need to make the connection by reading the Script output
> 
> Yeah, good question. If we can't show a specific label for each faker argument, an inline preview of the faker call would be good UI to show.

What if: if there are args, we just display the whole sample function call with args in a readonly text input? This is what I had in mind. Thoughts? @ncarbon @kpamaran

File: packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.spec.tsx (hunk 7 above)
Comment (ncarbon) [reply packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.spec.tsx:630]: Addressed

**Comment by kpamaran** [comment]
> What if: if there are args, we just display the whole sample function call with args in a readonly text input? This is what I had in mind. Thoughts?

I have confidence the sample call should be displayed. I'm not sure if it should be in an input, because the preview is not part of the form data (non-semantic usage). It could be some inline text above the field set. @jcobis @ncarbon

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx (hunk 8 above)
Comment (ncarbon) [reply packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: This is no longer applicable since we will show a preview of the faker call.

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx (hunk 10 above)
Comment (ncarbon) [reply packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: No longer applicable.

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 11]
```diff
@@ -0,0 +1,165 @@
+import { type Logger, mongoLogId } from '@mongodb-js/compass-logging/provider';
+import type { FakerArg } from './script-generation-utils';
+import { faker } from '@faker-js/faker/locale/en';
+
+const MAX_FAKER_ARGS_LENGTH = 10;
+const MAX_FAKER_STRING_LENGTH = 1000;
+const MAX_FAKER_ARGS_DEPTH = 3;
+const MAX_FAKER_NUMBER_SIZE = 1000;
```
Comment (ncarbon) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: Thoughts on these values? @jcobis

Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: I feel like we could bump `MAX_FAKER_NUMBER_SIZE` to `10000` at least

File: packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.spec.tsx (hunk 7 above)
Comment (ncarbon) [reply packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.spec.tsx:630]: Oops, I added the assertion in `utils.spec.ts`. Fixing.

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx [hunk 12]
```diff
@@ -24,16 +27,37 @@ const labelStyles = css({
   fontWeight: 600,
 });
 
+const parseFakerArg = (arg: FakerArg): string => {
```
Comment (kpamaran) [inline packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: nit: there's a `parseFakerArgs` with a different behavior. One should be renamed

**Review by kpamaran — APPROVED** [review]

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx (hunk 12 above)
Comment (gribnoysup) [reply packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: I can't find another function you mentioned, but the name still doesn't make too much sense: it seems to be the opposite of parse, takes the arguments and serializes them to string

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx [hunk 13]
```diff
@@ -24,16 +27,37 @@ const labelStyles = css({
...
+  if (typeof arg === 'object' && arg !== null && 'json' in arg) {
+    try {
+      return JSON.stringify(JSON.parse(arg.json));
+    } catch {
+      return '';
+    }
+  }
+  return arg.toString();
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: Based on the type definition, doesn't look like the array case is being handled well at all. If `arg` is `FakerArg[]` (which is part of the `FakerArg` type, it recursively references itself), you will just call prototype `toString` on it. Is this a mistake in how types are defined or in the method logic?

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+  return { moduleName, methodName, fakerModule };
+}
+
+function isAllowedHelper(moduleName: string, methodName: string) {
+  return (
+    moduleName !== 'helpers' ||
+    methodName === 'arrayElement' ||
+    methodName === 'arrayElements'
+  );
+}
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: Hard to understand the intent here reading this expression, also function name doesn't match what's going on here, we're not validating only the helpers, but all method names. I would suggest to restructure it to be more explicit about what we're actually checking:


```suggestion
function isAllowedFakerFn(moduleName: string, methodName: string) {
  if (moduleName !== 'helpers') {
    return true // all non-helper modules are allowed unconditionally
  }
  return methodName === 'arrayElement' || methodName === 'arrayElements' // only array helpers are allowed
}
```

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+ * ]
+ * becomes
+ * [ { a: 1 }, { b: 2 } ]
+ */
+function prepareFakerArgs(args: FakerArg[]) {
+  return args.map((arg) => {
+    if (typeof arg === 'object' && arg !== null && 'json' in arg) {
+      return JSON.parse((arg as { json: string }).json);
+    }
+    return arg;
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts:196]: `FakerArg` is recursively an array of the same type and this is not handled. Is this type actually defined correctly? There is a lot of code here now that uses `FakerArg[]`, but `FakerArg` type already includes an array, so either marking it as an array is redundant, or the type is wrong and it's always supporsed to be a flat array of arguments

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 16]
```diff
@@ -0,0 +1,189 @@
+import { type Logger, mongoLogId } from '@mongodb-js/compass-logging/provider';
+import type { FakerArg } from './script-generation-utils';
+import { faker } from '@faker-js/faker/locale/en';
+
+const MAX_FAKER_ARGS_LENGTH = 10;
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: Looking through the docs I can't see a case where it's ever more than 1 argument actually, any reason not to limit this more aggresively?

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+      if (!areFakerArgsValid(arg, depth + 1)) {
+        return false;
+      }
+    } else if (
+      typeof arg === 'object' &&
+      arg !== null &&
+      typeof (arg as { json?: unknown }).json === 'string'
+    ) {
+      try {
+        const parsedJson = JSON.parse((arg as { json: string }).json);
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: You can use an `in` check here which will also make it so that typescript know that property exists, so that you don't need to do extra assertions later:

```suggestion
      'json' in arg && typeof arg.json === 'string'
    ) {
      try {
        const parsedJson = JSON.parse(arg.json);
```

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+ *   { json: '{"a": 1}' },
+ *   { json: '{"b": 2}' },
+ * ]
+ * becomes
+ * [ { a: 1 }, { b: 2 } ]
+ */
+function prepareFakerArgs(args: FakerArg[]) {
+  return args.map((arg) => {
+    if (typeof arg === 'object' && arg !== null && 'json' in arg) {
+      return JSON.parse((arg as { json: string }).json);
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: Redundant assertion

```suggestion
      return JSON.parse(arg.json);
```

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx (hunk 12 above)
Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: Renaming to `stringifyFakerArg`

File: packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx (hunk 13 above)
Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/faker-mapping-selector.tsx]: Thank you, going to handle arrays recursively

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts (hunk 14 above)
Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: Refactored for readability

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts (hunk 15 above)
Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/utils.ts:196]: The recursive type definition is to support nested arrays.

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts (hunk 16 above)
Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: I see some examples of 2 arguments, such as `fromCharacters` and `arrayElements`, but none more than that. Will reduce this to 2.

Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/utils.ts]: Ah, I see the issue, this is just an imprecise variable name. Should be `MAX_ARRAY_LENGTH`, will rename. I can also add validation to ensure top-level args are limited to 2 at most.
